### DISCUSSION TITLE
AdGuardHome listens only on IPv6 on an IPv4-only machine, causing the service to fail

### DISCUSSION THREAD

**Discussion by ChenYun10** [AdguardTeam/AdGuardHome#7395] [Q&A] [answered]
System Version:
Debian11
Processor architecture:
amd64
Software version:
0.107.53
Problem Description:
During use, AdGuardHome was found to only listen to IPv6 and not to IPv4 connections when the environment is normal. As a result, the configured DoH and DoT cannot connect to the server at all. I have IPv6 at home, but I want to use IPv4 in my home area network Unencrypted DNS (udp53) really can't work.
Please answer, thank you.

**Comment by MNLierman** [comment, marked as answer]
Same. Did you find a solution? I guess if I find one, I'll try to remember to comment back here. It's odd that you had this issue recently, I wonder if there was a code change in AdGuardHome that broke some setups. My setup is a Docker container and I've removed and redeployed the container for the last 2 hours trying to figure this damn crap out and I'm so fed up. Honestly, I might just try PiHole and call it a day. I have no F-ing clue why AdGuard is acting like this.

→ [My Actual Solution](https://github.com/AdguardTeam/AdGuardHome/discussions/7395#discussioncomment-11666275)

**Reply by ChenYun10** [reply]
I had to reload, and I had no other way

**Reply by ChenYun10** [reply]
And I've had this kind of issue with multiple versions I've used before

**Reply by MNLierman** [reply]
I actually found the better solution, I'm posting it now.

**Reply by ChenYun10** [reply]
OK, thank you
Actually, I use this because I'm in China, but some ISPs in China have too much DNS pollution and hijacking😂

**Comment by MNLierman** [comment]
**MY SOLUTION AND ADVENTURE IN PORT BINDING**
(Proudly written without the help of AI – yes, I said without.)

I figured out what was happening for mine. Just for anyone who finds this issue. 

This happened on my office network, at the end I talk about extra info I did for my home network not specifically related to this AdGuardHome (AGH) issue.

**Problem:** AdGuardHome Docker container was not binding to IPv4 address and instead only binding to IPv6. Essentially, without giving any error, it was not binding to 0.0.0.0 and `netstat -tulpn | grep 53` on AdGuardHome shows only `:::53` and that was it.

**Discovery:** I had the docker container for adguard on a network with other containers so that I could manage things in that regard. While I wasn't initially seeing an error message about being unable to bind, since IPv4 was missing, I assumed it was happening, just with a silent continue. I believe that AGH is programmed with a silent continue function. If it cannot bind to a certain adapter or address, and you have selected “Bind All” then it will not complain.

**Solution:** From that assumption that AGH is silently unable to bind and continuing without error. Using `netstat -tulpn | grep 53` (or other port / app name you want or search for), I started looking at each container on the same network as AGH, and found a container that was set to scan and attempt to bind to everything within access, an aggressive feature I didn't need in my circumstance, as I only needed that particular app to bind locally on **`lo 127.0.0.1`**.

**Next**, that lead me to check the host computer, and on the host computer I found the same thing. Now I don't have my notes in front of me at the moment, but I believe it was dnsmasq on the host computer, which was set to also bind to everything possible, which includes the local bridge adapter for the AGH container. I changed dnsmasq settings to only bind to `lo` and `127.0.0.1`. After that, I changed the config file for AGH to explicitly bind to `0.0.0.0`, `127.0.0.1`, and 172.x.x.x (the bridge adapter). 

**Finally**, I found this last piece was important too, I found that if containers on a specific network moved around they would change IPs. I thought that Docker would set/prefer static IPs and try to keep containers on the same IP. Apparently not. So I explicitly set, using portainer, the IP address field for both ipv4 and IPv6 to use the specific IP that I wanted so that it containers moved around, Docker was restarted, or containers were turned off and on, that the IPs didn't jump around. This resolved my issue completely.

**Conclusion**, I believe redeploying AGH may fix other people's problems only if it's competitively binding. Ie, it's first to bind to an adapter or IP, and if the host is restarted, it may lose in that binding competition. The proper way is check for other things that are binding unnecessarily, since AGH is set to automatically continue on error. Use `netstat -tulpn | grep 53`(or other port you want to check for). I found other containers/apps that were unnecessarily binding to things I didn't need them to bind. You can also create a special network just for AGH to use. I called mine `app_net_agh1` since I am now prepping a second AGH container. Then, you can specify in the Docker run command, Docker file, or in Portainer the IP you want AGH to have. Finally I also edited the AGH config yaml file and explicitly specified the addresses to attempt to bind to.

**HOME NETWORK**
**Extra Info:** On my home network, I decided I wanted to create a macvlan for AGH, so I did that, but found the host cannot contact that AGH IP, never found why, I think it's a Docker bug (even checked IPTables). The easy fix there is to tell the host to use the IP of the bridge adapter (second network) that you also attach to AGH. You can have multiple networks attached to containers. So while the 10.5.5.x IP on my home network that AGH picked up was unable to be pinged by the host, it's local bridge IP can be pinged so I set that IP in the hosts `/etc/resolve.conf`. 

Now it's also important to know that many installations of Ubuntu and Debian now move that resolv.conf responsibility to `/etc/systemd/network.conf` (I think that's the one). On there you can specify a main DNS and a `fallbackdns=`. I put fallback to something else like NextDNS so if the host couldn't contact the AGH container, it would fallback its DNS to NextDNS and not lose internet connectivity, that way it could still resolve the internet world and I could still access it remotely if I had to.

**Reply by ChenYun10** [reply]
Thank you

**Comment by Generator** [comment]
Having this issue for some time, notice that no matter if i change something or not ipv4 stops listening requests.

Device IP is 10.0.0.40, Adguard Home running on Docker with network host, for some time that IP worked (even from system) without issues.
After some time without rebooting or changing anything, only update docker image.

10.0.0.40 stop working

```
10.0.0.40 fail
127.0.0.1 fail
::1       OK
```

After trying to figured out what's the issues without success, tried clean install on system using binary.

```
10.0.0.40 fail
127.0.0.1 OK
::1       fail
```

Tried the same clean install with docker

```
10.0.0.40 fail
127.0.0.1 OK
::1       fail
```

After changing some settings (CNAME, block lists...) and restart docker container, changed again

```
10.0.0.40 OK
127.0.0.1 fail
::1       fail
```

Going to try checking all settings over again, see what changes.

PS: testing inside the container gives same results

**Reply by ChenYun10** [reply]
All ports (443 DoH and 853 DoT/53 UDP/TCP) will only listen to IPv6 instead of IPv4 after a period of time, but the latest stable version does not have this problem.😂

**Reply by aersi** [reply]
I meet this too， AG version is v0.107.69。

**Comment by EricHarbers** [comment]
I use AdGuardHome v0.107.71 on a Raspberry Pi 5, using Ubuntu 24.04 server.
When i setup system using a dual stack ipv4/ipv6 configuration, and i want to use tls with DoT port 583, i see the system starting up for 1pv4 and connecting to port 0.0.0.0:853 without problem. After that, I see the system initiating the ipv6 part. When it tries to connect to port [::]:853 it returns with an error: address already in use, and stops adguard home to run.
After that i disabled the whole ipv6 part on my lan network, and pi, and now AdGuardHome runs without problems. It would be nice if adguardhome will work dual-stack on both 0.0.0.0:853 and [::]:853 without collision errors.

**Comment by MNLierman** [comment]
Hi,
Address already in use means that Docker processes themselves reserved that
port for that address and restarting AdGuard Home container itself doesn't
always fix this, as my testing and research has shown that this is a Docker
bug. (For that I've used docker stop, killall docker, and docker start
commands.) However, with the latest version of Docker beta and AdGuard Home
beta, I haven't had this issue in 10 months. I setup Docked and AdGuard and
outside any automatic updates, I haven't pressed any update buttons or done
anything different.

The comment I made almost a year ago, on this thread or a previous, about
ports – Docker, and AdGuard still work and the issue did not arise again. I
have had multiple power failures (15 minutes battery backup on Pis and
Nvidias which has run out at least 3 different times), and never had to
screw around with anything since then.

I do recall that killing the AdGaurd process inside the Docker image
doesn't work and I had to ensure that it was accessible as a individually
configured static IP using mac vlan. It's a simple process, you setup a
network on Docker and put in all your router DHCP information but give
AdGuard an IP without a address pool just outside your DHCP. For example,
my router issues addresses starting from x.x.x.100 - x.x.x.200 and then mac
vlan gives Docked containers addresses from x.x.x.220 - x x.x.230.

IPv6: Add a second network for IPv6 and specify a pool in Docked IPv6
config file. I used 2607:x:x:x:d0cc:4::1/64 but I have AdGuard configured
specifically for 2607:x:x:x:d0cc:4::ad1. This is done in the Portainer IPv6
static IP config box, and the IPv4 was also given here to ensure that it
doesn't change the address when it reboots. So the network and pool is
there in Networks just to give it context and subnet mask information but
you are supplying this same information into Portainer static info, so it
may actually grab 2 IPs of v4 and v6.

Finally, if you can't get IPv6 to work where the AdGuard Home is global
statically routable, which REQUIRES ISP /64 or /128 IPv6 static block, then
AdGuard will only pickup globally routable addresses within the assigned
pool, and those will change, but what matters is that it hits DNS servers
on IPv6.

Then, if this still doesn't work and your IPv6 is not globally routable and
working, ie. AdGuard Home can't ping IPv6.google.com or another IPv6
address, then it should still be able to request and serve IPv6 information
over IPv4 locally. This means when your computer/device asks the AdGuard
Home docker container on it's IPv4 address for an IPv6 DNS query, it will
be able to tell your querying dns what that IPv6 address is and it serves
that information on your local network through IPv4. That device will then
use the appropriate IPv6 address.

Also, I've had to “killall docker” before but do this after docker stop,
then follow with docker start and this frees up any reservations – usually.
A full restart of Ubuntu shouldn't required, but it could help with initial
configuration.

Just to reiterate tho, I have a fully working dual stack ADG on IPv4 and
IPv6 with globally routable IPv6 addresses able to route from outside my
network, through IPv6, to ADG and it serves DNS info on this. To note, I
did this by allowing IPv6 global IPs through my firewall and only allow
this for a whitelist set of /64s and /128s. This is done through -allow on
incoming and forward of iptables on your router.

I hope this is of some help, but if unclear, find my previous comment and
follow the information I posted previously.

**Reply by EricHarbers** [reply]
I don't use docker. I use Ubuntu 24.04 strait on a raspberry pi 5. I've checked port usage by other processes, but nothing used that port. I also tried setting the port to a unusual number, eg 8853, and that produces the same error. Is seems that the port listening will be used for ipv4 and ipv6 for exactly the same port. If i use only ipv4, the problem is away.

**Reply by EricHarbers** [reply]
I’m running AdGuard Home v0.107.71 natively on Ubuntu 24.04 on a Raspberry Pi 5, not in Docker.

I observed the following issue:

When using dual-stack IPv4/IPv6, and configuring the same DoT port (e.g., 853) for both IPv4 and IPv6, AdGuard fails to start the IPv6 listener with the error address already in use.

Changing to an unusual port (e.g., 8853) produces the same behavior.

Running IPv4 only works fine, and AdGuard starts without errors.

No other processes are using the port (netstat/ss checks confirm this).

This strongly suggests that the issue is internal to AdGuard Home: binding the same port for IPv4 and IPv6 fails, even though the addresses are distinct (0.0.0.0 vs [::]).

It would be great if AdGuard Home could correctly bind a single port on both IPv4 and IPv6 in a dual-stack environment, so dual-stack DNS-over-TLS/HTTPS configurations work without this collision.
